commands/filescommands/objectcommands/uploaddir.go

```diff
@@ -0,0 +1,235 @@
+	"github.com/jrperritt/rack/internal/github.com/dustin/go-humanize"
+	"github.com/jrperritt/rack/internal/github.com/rackspace/gophercloud/openstack/objectstorage/v1/objects"
+	"github.com/jrperritt/rack/util"
+)
+
+var uploadDir = cli.Command{
+	Name:        "upload-dir",
+	Usage:       util.Usage(commandPrefix, "upload-dir", "--container <containerName> --dir <dirName>"),
+	Description: "Uploads the contents of a local directory to a container",
+	Action:      actionUploadDir,
+	Flags:       util.CommandFlags(flagsUploadDir, keysUploadDir),
+	BashComplete: func(c *cli.Context) {
+		util.CompleteFlags(util.CommandFlags(flagsUploadDir, keysUploadDir))
+	},
+}
+
+func flagsUploadDir() []cli.Flag {
+	return []cli.Flag{
+		cli.StringFlag{
+			Name:  "container",
+			Usage: "[required] The name of the container to upload the object upload",
+		},
+		cli.StringFlag{
+			Name:  "dir",
+			Usage: "[required] The name the local directory which will be uploaded",
+		},
+		cli.StringFlag{
+			Name:  "content-type",
+			Usage: "[optional] The Content-Type header that will be set on all objects.",
+		},
+		cli.IntFlag{
+			Name:  "content-length",
+			Usage: "[optional] The Content-Length header that will be set on all objects.",
+		},
+		cli.StringFlag{
+			Name:  "content-encoding",
+			Usage: "[optional] The Content-Encoding header that will be set on all objects. By default, the uploaded content will be gzipped.",
+		},
+		cli.IntFlag{
+			Name:  "concurrency",
+			Usage: "[optional] The amount of concurrent workers that will upload the directory.",
+		},
+		cli.BoolFlag{
+			Name:  "quiet",
+			Usage: "[optional] By default every file upload will be outputted. If --quiet is provided, only a final summary will be outputted.",
+		},
+	}
+}
+
+var keysUploadDir = []string{}
+
+type paramsUploadDir struct {
+	container   string
+	dir         string
+	stream      io.ReadSeeker
+	opts        objects.CreateOpts
+	concurrency int
+	quiet       bool
+}
+
+type commandUploadDir handler.Command
+
+func actionUploadDir(c *cli.Context) {
+	command := &commandUploadDir{
+		Ctx: &handler.Context{
+			CLIContext: c,
+		},
+	}
+	handler.Handle(command)
+}
+
+func (command *commandUploadDir) Context() *handler.Context {
+	return command.Ctx
+}
+
+func (command *commandUploadDir) Keys() []string {
+	return keysUploadDir
+}
+
+func (command *commandUploadDir) ServiceClientType() string {
+	return serviceClientType
+}
+
+func (command *commandUploadDir) HandleFlags(resource *handler.Resource) error {
+	if err := command.Ctx.CheckFlagsSet([]string{"container", "dir"}); err != nil {
+		return err
+	}
+
+	c := command.Ctx.CLIContext
+
+	opts := objects.CreateOpts{
+		ContentLength: int64(c.Int("content-length")),
+		ContentType:   c.String("content-type"),
+	}
+
+	if c.IsSet("content-encoding") && c.String("content-encoding") != "gzip" {
+		opts.ContentEncoding = c.String("content-encoding")
+	}
+
+	conc := c.Int("concurrency")
+	if conc <= 0 {
+		conc = 5
```

Comment:
Is this the number of goroutines that get spawned? If so, I wouldn't have it as a user-defined flag. Instead, I'd create a pool of workers the size of the maximum-allowed concurrent connections to Cloud Files (if such a thing exists) and continually reuse the workers until there aren't any remaining files to upload.